### PR TITLE
Suggestion: remove stream use from strategy orchestration

### PR DESCRIPTION
**Pull request by Farenheith**
I suggest to not use stream when orchestrating the serialization of protobuf or json.

The reason is that, either protobuf, fast-json or json serialization doesn't return a stream, but actually string or Unit8Array.
The only component that returns a stream is the gzip.
 
This detail can produce problems with the protobuf and json transforms, because they run deserialize operations over chunks received from gzip (when the serialization use gzip), and it's possible that the chunk is not the whole information, thus, the deserialization may throw an error or function incorrectly.

I think the best approach, in this case, is to treat gzip as a whole new Serialization Strategy and to give the possibility for the user, in the Serializer class, to compose chained serializers.

The first serializer may define the type of the input (if the type is specified), the following one, though, must receive a Serialized information and returns a Serialized information too, so, the contract of Serializer keep unchanged.

The use of Serializer without gzip may be something like this:

```ts
		const proto = new ProtobufStrategy<MyType>({
			attribute: 'a.b.Foo',
			proto: './foo.proto',
		});
		const serializer = new Serializer(proto);
		const write = await serializer.serialize(req);
		const read = await serializer.deserialize(write);
```

While the use with gzip may be something like this:

```ts
		const proto = new ProtobufStrategy<typeof req>({
			attribute: 'a.b.Foo',
			proto: './foo.proto',
		});
		const serializer = new Serializer(proto, new GzipStrategy());
		const write = await serializer.serialize(req);
		const read = await serializer.deserialize(write);
```

This approach also offers the advantage to specify the parametrization of gzip. Example:

```ts
		const proto = new ProtobufStrategy<typeof req>({
			attribute: 'a.b.Foo',
			proto: './foo.proto',
		});
		const serializer = new Serializer(proto, new GzipStrategy({
                   level: 9,
                }));
		const write = await serializer.serialize(req);
		const read = await serializer.deserialize(write);
```

Another advantage is that this contract creates a relation between the serialize and deserialize methods: both will use the exact same strategy, preventing misuses like serialize with a Strategy and deserialize with another.

I made the changes and I ask you to take a look in then, to decide if this proposal makes sense.

### CODE REVIEW

File: tsconfig.json [hunk 1]
```diff
@@ -12,7 +12,7 @@
     "experimentalDecorators": true,
     "importsNotUsedAsValues":"remove",
     "target": "es2018",
-    "sourceMap": false,
+    "sourceMap": true,
```
Comment (Farenheith) [inline tsconfig.json:15]: I moved the sourceMap false to the tsconfig.build.json. This way, the dist still doesn't have the .map's, but the debug experience is better

File: src/strategy/protobuf/index.ts [hunk 2]
```diff
@@ -1,66 +1,35 @@
...
-					throw err;
-				}
-				resolve(root?.lookupType(options.attribute));
-			});
-		});
+		const type: Type | undefined = await new Promise(async (resolve, reject) =>
+			load(options.proto, (err: Error | null, root?: Root) =>
+				err ? reject(err) : resolve(root?.lookupType(options.attribute)),
+			),
+		);
```
Comment (Farenheith) [inline src/strategy/protobuf/index.ts:20]: The throw err would be an untreated error and could shut down the service. Using reject instead gives an opportunity for the user to catch and treat the error if he wants to

File: src/strategy/json/types/index.ts [hunk 3]
```diff
@@ -1 +1,3 @@
+/* istanbul ignore file */
```
Comment (Farenheith) [inline src/strategy/json/types/index.ts:1]: ts-jest has an issue where interface files are considered in the Istanbul report. This comment makes Istanbul ignore this file

File: src/strategy/json/index.ts [hunk 4]
```diff
@@ -1,46 +1,31 @@
...
-		return transform;
+	async deserialize<T extends A>(
+		content: Serialized,
+		encode: BufferEncoding = 'utf-8',
+	): Promise<T> {
+		return JSON.parse(
+			typeof content === 'string'
+				? content
+				: Buffer.from(content).toString(encode),
+		);
```
Comment (Farenheith) [inline src/strategy/json/index.ts:29]: for both, the json strategy and the protobuf strategy, there is no need anymore of the transform classes, as no stream is being manipulated, so, the logic of the strategy and the transform has been merged in both cases

File: test/unit/serializer/index.spec.ts [hunk 5]
```diff
@@ -34,13 +56,58 @@ describe('index.ts', () => {
...
+
+	it('should work with json + gzip', async () => {
+		const req = {
+			bar: 'abc',
+		};
+		const strategy = new JsonStrategy<typeof req>();
+		const serializer = new Serializer(strategy, new GzipStrategy());
+
+		const write = await serializer.serialize(req);
+		const read = await serializer.deserialize(write);
```
Comment (Farenheith) [inline test/unit/serializer/index.spec.ts:163]: The test cases I wrote here can be used even if this suggestion is not accepted with minor adjustments

File: src/strategy/serializer.ts [hunk 6]
```diff
@@ -1,17 +1,17 @@
...
-		content: Stream,
-		options?: StrategyOptions,
-	): Promise<Stream> | Stream;
-	deserialize(
-		content: Stream,
-		options?: StrategyOptions,
-	): Promise<StrategyStream> | StrategyStream;
+export type Serialized = string | ArrayBuffer | Uint8Array;
+
+export interface SerializerStrategy<I, O extends Serialized | Stream> {
```
Comment (Farenheith) [inline src/strategy/serializer.ts:11]: Notice that a SerializerStrategy can still return a stream (as the case with gzip). But by the overall default implementation streams will only be used in the chained strategies (where the only default chaining strategy is gzip)

**Review by gustavo-rodrigues-dev — APPROVED** [review]

**Review by danielgalleni — APPROVED** [review]

**Pull request merged**